2022/10/ssh-askpass-on-Mac-OS-for-agent-confirmation.md

```diff
@@ -0,0 +1,91 @@
+---
+author: "Bharathi Ponnusamy"
+date: 2022-10-26
+title: "Ssh-askpass on Mac OS for the agent confirmation"
+github_issue_number: 
+tags:
+- ssh 
+- ssh-askpass
+- mac
+- security
+---
+
+![On a summer evening](/blog/2022/10/ssh-askpass-on-Mac-OS-for-agent-confirmation/banner.jpg)
+[Photo](https://flic.kr/p/2nUPsJQ) by [Kristoffer Trolle](https://www.flickr.com/people/kristoffer-trolle/), [CC BY 2.0](https://creativecommons.org/licenses/by/2.0/)
+
+At End Point Dev we mostly use the SSH keys for authentication when connecting to remote servers and Git services. The majority of the time, the servers we are trying to visit are barred from direct access and require a middle "jump server" instead.
+
+Enabling SSH Agent Forwarding makes it easier to reuse the SSH private keys. It keeps the private keys on our local machine to authenticate with each server in the chain without entering a password.
+
+However, this approach comes with an inherent risk where the agent can be hijacked. This means a bad guy could use the SSH keys to compromise downstream servers.
+
+In this post, we’ll cover a simple way to protect against SSH Agent Hijacking. Also, we will see in detail how to configure a system-wide agent that runs under GUI for agent confirmation using ssh-askpass.
+
+## ssh-askpass - Agent Confirmation
+
+![agent confirmation dialog](/blog/2022/10/ssh-askpass-on-Mac-OS-for-agent-confirmation/ssh-askpass.png)
+
+It is strongly recommended to use the `-c` option on `ssh-add` command when adding your ssh keys to the ssh-agent in order to protect yourself against SSH Agent Hijacking.
+
+Agent forwarding has some inherent risk, however, this risk may be significantly reduced by using this `-c` option.
+
+When utilizing the ssh-askpass, a prompt will show on your local computer to confirm the usage of the key each time a request is made to use the private key loaded in the SSH agent. Doing this makes it harder for an attacker to utilize the private key without permission.
+
+If case if you don’t want to use agent confirmation, then it’s best to use ProxyJump instead of agent forwarding.
+
+
+### Install ssh-askpass with Homebrew
+```
+$ brew tap theseal/ssh-askpass
+$ brew install ssh-askpass
+```
+You might see some warnings. Go ahead and proceed with them.
+
+### Install ssh-askpass with MacPorts
+```
+$ sudo port install ssh-askpass
+```
+
+### Install ssh-askpass from Source Code
+```
+$ cp ssh-askpass /usr/local/bin/
+$ cp ssh-askpass.plist ~/Library/LaunchAgents/
+$ launchctl load -w ~/Library/LaunchAgents/ssh-askpass.plist
+```
+
+## To start ssh-askpass now and restart at login
+
+```
+$ brew services start theseal/ssh-askpass/ssh-askpass
```

Comment:
@bharathi3p ssh-askpass is not actually a service; there is nothing to "start" or "restart" in a traditional sense.

Run `brew services list | grep ^ssh-askpass` and look at the `plist` file it shows there.

At least for me, that runs a shell script that sets the `SSH_ASKPASS` and `SUDO_ASKPASS` environment variables and stops `ssh-agent` so that it has to pick up these environment variables when it restarts.

So homebrew calls it a service, but I think it's worth noting here that we're not really "starting" anything, because `ssh-askpass` is just a normal program that runs on demand by ssh-agent itself.

Maybe that seems too picky, too fine of a difference, but I think it's important since anyone looking to find `ssh-askpass` in their process list with `ps` will be disappointed, and this is why.

Would you mind maybe throwing in something about this so it doesn't sound like a running daemon itself?

Reply:
@jonjensen that's right, I Just made the change.